### PR TITLE
Proper support for cross-crate recursive const stability checks

### PR DESCRIPTION
**Pull request by RalfJung**
~~Stacked on top of https://github.com/rust-lang/rust/pull/132492; only the last three commits are new.~~

In a crate without `staged_api` but with `-Zforce-unstable-if-unmarked`, we now subject all functions marked with `#[rustc_const_stable_indirect]` to recursive const stability checks. We require an opt-in so that by default, a crate can be built with `-Zforce-unstable-if-unmarked` and use nightly features as usual. This property is recorded in the crate metadata so when a `staged_api` crate calls such a function, it sees the `#[rustc_const_stable_indirect]` and allows it to be exposed on stable. This, finally, will let us expose `const fn` from hashbrown on stable.

The second commit makes const stability more like regular stability: via `check_missing_const_stability`, we ensure that all publicly reachable functions have a const stability attribute -- both in  `staged_api` crates and `-Zforce-unstable-if-unmarked` crates. To achieve this, we move around the stability computation so that const stability is computed after regular stability is done. This lets us access the final result of the regular stability computation, which we use so that `const fn` can inherit the regular stability (but only if that is "unstable"). Fortunately, this lets us get rid of an `Option` in `ConstStability`.

This is the last PR that I have planned in this series.

r? @compiler-errors

### CODE REVIEW

File: tests/ui/consts/rustc-const-stability-require-const.rs [hunk 1]
```diff
@@ -56,9 +56,3 @@ const fn barfoo_unmarked() {}
 #[rustc_const_stable(feature = "barfoo_const", since = "1.0.0")]
 pub const fn barfoo_unstable() {}
 //~^ ERROR can only be applied to functions that are declared `#[stable]`
-
-// `#[rustc_const_stable_indirect]` also requires a const fn
-#[rustc_const_stable_indirect]
-#[unstable(feature = "unstable", issue = "none")]
-pub fn not_a_const_fn() {}
```
Comment (RalfJung) [inline tests/ui/consts/rustc-const-stability-require-const.rs:63]: This regresses complaining about rustc_const_stable_indirect on non-const-fn... but keeping that error would make the code even more messy, and we also don't error for rustc_promotable on non-const-fn, so :shrug: 

There's an ongoing effort to redo the way we do attributes, that should hopefully make this easier in the future.

File: tests/ui/traits/const-traits/staged-api.rs [hunk 2]
```diff
@@ -37,8 +37,7 @@ const fn const_context() {
     //[stable]~^^ ERROR not yet stable as a const fn
     Foo::func();
     //[unstable]~^ ERROR cannot use `#[feature(local_feature)]`
-    //[stable]~^^ cannot be (indirectly) exposed to stable
-    // We get the error on `stable` since this is a trait function.
+    // FIXME: in [stable] we should still be complaining about this being a trait call.
```
Comment (RalfJung) [inline tests/ui/traits/const-traits/staged-api.rs:40]: Trait calls don't show any errors any more... but the old trait call checks are broken anyway so I figured it's fine to break them even more, and wait until @compiler-errors lands the new checks. :)

**Comment by RalfJung** [comment]
That's a tricky one... when one force-unstable-if-unmarked crate calls things from another one, how does that work? The callee crate has everything under the `rustc_private` feature gate, which the calling crate does not have enabled. Somehow this works for regular stability; we'll have to use the same logic for const stability.

File: tests/ui/traits/const-traits/staged-api.rs (hunk 2 above)
Comment (RalfJung) [reply tests/ui/traits/const-traits/staged-api.rs:39]: See [here](https://github.com/rust-lang/rust/pull/132479#issuecomment-2454641534) for how I think this check should work.

I can add that to this PR, if you prefer.

Comment (RalfJung) [reply tests/ui/traits/const-traits/staged-api.rs]: After rebasing this PR no longer affects the const-trait tests. :)

File: compiler/rustc_const_eval/src/check_consts/mod.rs [hunk 3]
```diff
@@ -53,10 +53,11 @@ impl<'mir, 'tcx> ConstCx<'mir, 'tcx> {
     }
 
     pub fn enforce_recursive_const_stability(&self) -> bool {
-        // We can skip this if `staged_api` is not enabled, since in such crates
-        // `lookup_const_stability` will always be `None`.
+        // We can skip this if neither `staged_api` nor `-Zforrce-unstable-if-unmarked` are enabled,
```
Comment (compiler-errors) [inline compiler/rustc_const_eval/src/check_consts/mod.rs:56]: typo on the arg

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
r=me after really tiny nit

**Pull request merged**